Fetches control group resource information

tylersmalley opened the pull request:
These are made available to the status API `api/status` and `kbnServer.metrics`.

jbudz commented:
Test failure looks legitimate, I'm guessing we need to wait for the promise to resolve on previous status tests

jbudz commented on src/server/status/cgroup.js:
```diff
@@ -0,0 +1,136 @@
+
+// Logic from elasticsearch/core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java
+
+const CONTROL_GROUP_RE = new RegExp('\\d+:([^:,]+(?:,[^:,]+)?):(/.*)');
```
do we need to update this to match? https://github.com/elastic/elasticsearch/commit/00a8b8799f6b7de803bc8afeaf7de392f3d0d28c

tylersmalley commented on src/server/status/metrics.js:
```diff
@@ -1,40 +1,49 @@
   const port = config.get('server.port');
   const timestamp = new Date().toISOString();
-  return {
+  const cgroup = await getAllStats();
```
Now preventing lookups when we determine there is no cgroups information. https://github.com/elastic/kibana/pull/10402/commits/655f54bc3b113f08c88c5ee24d3e3c4a00125dcc

tylersmalley commented:
@tsullivan - I pushed a fix for the docker cgroup path.

Here is a docker image on Dockerhub to help with testing:

[tylersmalley/kibana:6.0.0-alpha1-SNAPSHOT-1ae1a6a
](https://hub.docker.com/r/tylersmalley/kibana/)

You can limit the CPU to exposed throttled metrics by passing `--cpus=0.1` to `docker run`

And here are snapshots of the PR:
```
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-1ae1a6a/kibana/kibana-6.0.0-alpha1-SNAPSHOT-amd64.deb
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-1ae1a6a/kibana/kibana-6.0.0-alpha1-SNAPSHOT-darwin-x86_64.tar.gz
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-1ae1a6a/kibana/kibana-6.0.0-alpha1-SNAPSHOT-i386.deb
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-1ae1a6a/kibana/kibana-6.0.0-alpha1-SNAPSHOT-i686.rpm
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-1ae1a6a/kibana/kibana-6.0.0-alpha1-SNAPSHOT-linux-x86.tar.gz
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-1ae1a6a/kibana/kibana-6.0.0-alpha1-SNAPSHOT-linux-x86_64.tar.gz
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-1ae1a6a/kibana/kibana-6.0.0-alpha1-SNAPSHOT-windows-x86.zip
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-1ae1a6a/kibana/kibana-6.0.0-alpha1-SNAPSHOT-x86_64.rpm
```

When running, you can apply cgroup constraints to a process/group. To do so, run `systemd-cgls` to find the user slice which contains kibana. You can then use it to set CPUQuota (example: `systemctl set-property user.slice CPUQuota=100%`)

tsullivan commented on src/server/status/cgroup.js:
```diff
@@ -0,0 +1,138 @@
+
+function fileContentsToInteger(path) {
+  return readFile(path).then(data => {
+    return parseInt(data.toString());
```
I would recommend using the radix parameter of parseInt, because without it behavior could be unpredictable. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

tsullivan commented on src/server/status/cgroup.js:
```diff
@@ -0,0 +1,138 @@
+
+        switch(fields[0]) {
+          case 'nr_periods':
+            stat.number_of_elapsed_periods = parseInt(fields[1]);
```
radix parameter on these parseInts

tsullivan commented:
Thanks for providing the Docker images to test with - that was extremely helpful. 

I left just one comment about something that popped up in a few places. I'm surprised that radix for parseInt thing isn't part of our lint rules 😺

jbudz commented on src/server/status/metrics.js:
```diff
@@ -1,39 +1,67 @@
+    }
+
+    const cgroup = await cGroupStats({
+      cpuPath: config.get('cpu.cgroup.path.override') ? '/' : undefined,
```
do you think there's any need for the path to be configurable?  ie instead of a bool, cpu.cgroup.path = Joi.string()

tylersmalley replied:
These configuration options align with logstash/ES outlined here: https://github.com/elastic/logstash/issues/6797#issuecomment-286605180

jbudz replied:
👍  for consistency.  it's not clear to me from the thread if override is meant to be a bool, the current (eventually deprecated) implementation in es looks like a string.  if i'm missing something feel free to ignore.

tylersmalley replied:
You're right, good find. https://github.com/elastic/elasticsearch-docker/blob/master/build/elasticsearch/bin/es-docker#L37

Will update

jbudz commented:
Rebasing should fix the test failures

jbudz commented on src/server/status/cgroup.js:
```diff
@@ -0,0 +1,138 @@
+      return null;
+    }
+
+    throw err;
```
how would you feel about logging this error instead of throwing?  if something comes up we can still return non c-group stats

tylersmalley commented:
Addressed feedback and pushed new builds for testing:

Docker: `docker run -p 5601:5601 -e "ELASTICSEARCH_URL=http://172.16.0.63:9200" --cpus=0.1 tylersmalley/kibana:6.0.0-alpha1-SNAPSHOT-5812d7b`

```
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-5812d7b/kibana/kibana-6.0.0-alpha1-SNAPSHOT-amd64.deb
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-5812d7b/kibana/kibana-6.0.0-alpha1-SNAPSHOT-darwin-x86_64.tar.gz
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-5812d7b/kibana/kibana-6.0.0-alpha1-SNAPSHOT-i386.deb
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-5812d7b/kibana/kibana-6.0.0-alpha1-SNAPSHOT-i686.rpm
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-5812d7b/kibana/kibana-6.0.0-alpha1-SNAPSHOT-linux-x86.tar.gz
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-5812d7b/kibana/kibana-6.0.0-alpha1-SNAPSHOT-linux-x86_64.tar.gz
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-5812d7b/kibana/kibana-6.0.0-alpha1-SNAPSHOT-windows-x86.zip
https://download.elastic.co/kibana/staging/6.0.0-alpha1-SNAPSHOT-5812d7b/kibana/kibana-6.0.0-alpha1-SNAPSHOT-x86_64.rpm
```

jbudz commented on src/server/status/metrics.js:
```diff
@@ -1,39 +1,71 @@
-export function getMetrics({ event, config }) {
+export async function getMetrics(event, config, server) {
   const port = config.get('server.port');
   const timestamp = new Date().toISOString();
```
do you think we should set timestamp after cgroup stats have been read?  along the same lines, do you think we should only getMetrics once the previous getMetrics call is done so they don't queue up for frequent ops events?

tylersmalley replied:
Having it right at the ops event is probably best as it's actually involved in aggregations (request count, etc). Capturing cgroups is also a very quick operation in that it's reading most of the files in parallel and they aren't actually on disk. We should revisit when we add CPU percentage.

tylersmalley commented:
@jbudz documented configuration options - placed near existing status config.

![screenshot 2017-04-05 13 00 38](https://cloud.githubusercontent.com/assets/40265/24724504/2eafce6e-1a00-11e7-8e86-a28c7e096359.png)

tylersmalley commented:
@tsullivan - mind giving another look?

tylersmalley commented on src/server/config/schema.js:
```diff
@@ -27,6 +27,21 @@ module.exports = () => Joi.object({
 
+  cpu: Joi.object({
+    cgroup: Joi.object({
+      path: Joi.object({
```
The name was discussed [here](https://github.com/elastic/logstash/issues/6797#issuecomment-286605180) for Logstash/ES. While I agree we could omit "override" it maintains consistency with the other projects.

pickypg replied:
Ah that's unfortunate, but I guess it's intentional to really show that you're going against defaults.

tylersmalley commented:
5.x: https://github.com/elastic/kibana/pull/11163
